ctapipe/image/extractor.py

```diff
@@ -1287,18 +1289,127 @@ def __call__(
             is_valid=is_valid,
         )
 
-
 @lru_cache
-def exponential_scale_from_pulse_shape(camera: CameraDescription):
-    pulse_shape = camera.readout.reference_pulse_shape
-    sample_width = camera.readout.reference_pulse_sample_width
+def deconvolution_parameters(camera: CameraDescription, upsampling: int, window_width: int, window_shift: int):
+    assert(upsampling > 0)
+    assert(window_width > 0)
+
+    ref_pulse_shapes = camera.readout.reference_pulse_shape
+    ref_sample_width = camera.readout.reference_pulse_sample_width.to_value("ns")
+    camera_sample_width = 1.0 / camera.readout.sampling_rate.to_value("GHz")
+
+    assert(camera_sample_width >= ref_sample_width)
```

Comment:
That's asserting a property of the camera in something that has no control at all over the `CameraReadout` content, which I think is a bit strange. If that case should never happen, we should test this elsewhere.

If this case can happen in general, but this method needs it not to happen, again as above, raise a nice `ValueError` instead of asserting.

Reply:
> That's asserting a property of the camera in something that has no control at all over the `CameraReadout` content, which I think is a bit strange.

Are checks of input parameters (=safe programming) not wanted in ctapipe? If so I can remove those checks.

> If that case should never happen, we should test this elsewhere.

Not sure where - as I can't even find the location of the reference waveforms. Could someone more knowledgable do this?

> If this case can happen in general, but this method needs it not to happen, again as above, raise a nice ValueError instead of asserting.

Will do.

Reply:
> Not sure where - as I can't even find the location of the reference waveforms. Could someone more knowledgable do this?

These are (at least for simtel files) directly stored in the camera configuration part of the files.

> Are checks of input parameters (=safe programming) not wanted in ctapipe? If so I can remove those checks.

They are wanted in many cases, but then they shouldn't be `assert` statements but the corresponding exception, e.g. `TypeError` or `ValueError` with nice error messages.

Reply:
> > Not sure where - as I can't even find the location of the reference waveforms. Could someone more knowledgable do this?
> 
> These are (at least for simtel files) directly stored in the camera configuration part of the files.

I'm still not sure where to put such a check - my comprehension of this codebase is lacking, sorry. :/

Reply:
If I understand correctly, you check that the single pe pulse is sampled more finely than the sampling frequency of the actual data, correct?

Reply:
At least as fine as data; yes.

Reply:
Is that a general requirement or specific to this method?

Reply:
This in any case needs a very clear error message, because at least in the case of sim_telarray input files, the user's only way to fix this is to resimulate with a finer binned single pe pulse, it's completely outside the control of ctapipe.

Reply:
I'd say it's a general requirement - doesn't make sense to use a too coarse impulse response for simulations.

I did this - ok?
```
    if camera_sample_width < ref_sample_width:
        raise ValueError(
            f"ref_sample_width (got {ref_sample_width}) must be equal to or shorter than camera_sample_width (got {camera_sample_width}); need a reference single p.e. pulse shape with finer sampling!"
        )
```

Reply:
yes, looks good, thanks